### PR TITLE
expose ILoggerFactoryBuilder to WebHostSettings; adding app insights http tests

### PR DESCRIPTION
**Pull request by brettsam**
This changes how we get the `ILoggerFactoryBuilder`. It now allows you to pass one into the `WebHostSettings`, which should allow the CLI to properly write ILogger logs to the console. That work is tracked with https://github.com/Azure/azure-functions-cli/issues/130. This doesn't fix it, but opens things up for a fix.

It also allows for full HTTP tests that log directly to a mock App Insights channel for easy validation. I've added some comprehensive tests for a few core scenarios to make sure they continue working as we approach GA.

### CODE REVIEW

**Comment by brettsam** [comment]
@mamaso -- these are the HTTP ILogger tests I mentioned.
@fabiocav -- this allows passing in an `ILoggerFactoryBuilder` (but just for full framework; we'll do it differently in core like we discussed)

File: test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsTestFixture.cs [hunk 1]
```diff
@@ -1,40 +1,150 @@
...
+using Microsoft.Azure.WebJobs.Logging.ApplicationInsights;
 using Microsoft.Azure.WebJobs.Script.Config;
+using Microsoft.Azure.WebJobs.Script.WebHost;
+using Microsoft.Extensions.Logging;
 using Microsoft.WebJobs.Script.Tests;
 
 namespace Microsoft.Azure.WebJobs.Script.Tests.ApplicationInsights
 {
-    public abstract class ApplicationInsightsTestFixture : EndToEndTestFixture
+    public abstract class ApplicationInsightsTestFixture
```
Comment (brettsam) [inline test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsTestFixture.cs]: I switched all of these tests to run on an HttpServer. The `Manual` tests just call via the admin API to invoke the test. I did that to limit the number of fixtures (and thus hosts) that I had to create.

File: test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsTestFixture.cs [hunk 2]
```diff
@@ -1,40 +1,150 @@
...
-                c.Functions = new[] { "Scenarios" };
+                // turn this off as it makes validation tough
+                config.HostConfig.Aggregator.IsEnabled = false;
+
+                // Overwrite the generated function whitelist to only include two functions.
+                c.Functions = new[] { "Scenarios", "HttpTrigger-Scenarios" };
             };
         }
+
+        private void WaitForHost()
```
Comment (mamaso) [inline test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsTestFixture.cs]: nit: I feel like I've seen these in other fixtures - can we move them to a utility?

File: test/WebJobs.Script.Tests.Integration/TestScripts/Node/Scenarios/index.js [hunk 3]
```diff
@@ -88,8 +87,8 @@ var assert = require('assert');
     else if (scenario === 'appInsights') {
 
         var logPayload = {
-            InvocationId: context.executionContext.invocationId,
-            Trace: input.value
+            invocationId: context.executionContext.invocationId,
+            trace: input.value
         };
 
         context.log(JSON.stringify(logPayload));
```
Comment (mamaso) [inline test/WebJobs.Script.Tests.Integration/TestScripts/Node/Scenarios/index.js]: out of curiosity, does this pass if you don't stringify the log payload? iirc we're doing that under the covers

File: src/WebJobs.Script.WebHost/App_Start/WebHostSettings.cs [hunk 4]
```diff
@@ -31,6 +31,8 @@ public class WebHostSettings
 
         public TraceWriter TraceWriter { get; set; }
 
+        public ILoggerFactoryBuilder LoggerFactoryBuilder { get; set; } = new DefaultLoggerFactoryBuilder();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/App_Start/WebHostSettings.cs]: Oh, what have they done to the language. Surely this can't compile :) @fabiocav

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 5]
```diff
@@ -791,10 +791,7 @@ private void ConfigureDefaultLoggerFactory()
             scriptConfig.HostConfig.LoggerFactory.AddProvider(new FileLoggerProvider(traceWriteFactory,
                 (category, level) => (category == LogCategories.Function) && isFileLoggingEnabled()));
 
-            // Allow a way to plug in custom LoggerProviders.
-            ILoggerFactoryBuilder builder = scriptConfig.HostConfig.GetService<ILoggerFactoryBuilder>() ??
-                new DefaultLoggerFactoryBuilder();
-            builder.AddLoggerProviders(scriptConfig.HostConfig.LoggerFactory, scriptConfig, settingsManager);
+            scriptConfig.LoggerFactoryBuilder.AddLoggerProviders(scriptConfig.HostConfig.LoggerFactory, scriptConfig, settingsManager);
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: So LoggerFactory extensibility is in core SDK, but factory builder extensibility is only in Script? Why is that out of curiosity? What if someone (e.g. CLI) wants to plug in both - it does so in two different locations?

File: src/WebJobs.Script/Host/ScriptHostManager.cs [hunk 6]
```diff
@@ -128,7 +128,8 @@ public void RunAndBlock(CancellationToken cancellationToken = default(Cancellati
                     // Create a new host config, but keep the host id from existing one
                     _config.HostConfig = new JobHostConfiguration
                     {
-                        HostId = _config.HostConfig.HostId
+                        HostId = _config.HostConfig.HostId,
+                        LoggerFactory = new LoggerFactory()
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHostManager.cs]: What if someone had already set _config.LoggerFactory?

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 5 above)
Comment (brettsam) [reply src/WebJobs.Script/Host/ScriptHost.cs]: Script needs this `Builder` b/c we have to construct a `LoggerFactory` over-and-over again whenever the host restarts (since it's built based on host.json). In SDK, there's no real need for that. You can just build the LoggerFactory on your own and pass it to us -- you only ever need to do it once.

Comment (brettsam) [reply src/WebJobs.Script/Host/ScriptHost.cs]: Sorry, didn't answer your other question -- CLI will pass us in a custom `Builder` with console logging and we'll use that to construct the `LoggerFactory`.

File: src/WebJobs.Script/Host/ScriptHostManager.cs (hunk 6 above)
Comment (brettsam) [reply src/WebJobs.Script/Host/ScriptHostManager.cs]: There is no property `ScriptHostConfiguration.LoggerFactory` -- that lives in `JobHostConfiguration`, which we construct new here.

BTW -- @fabiocav and I have discussed all these config lifetimes for v2 as it can get pretty confusing what lives from host-to-host and what is created new each time. We want to make sure to address it in Core.

Comment (mathewc) [reply src/WebJobs.Script/Host/ScriptHostManager.cs]: Yeah, I meant what if they set _config.HostConfig.LoggerFactory.

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 5 above)
Comment (mathewc) [reply src/WebJobs.Script/Host/ScriptHost.cs]: I do find all these builders/factories hard to understand, but likely I don't have all the context.

File: src/WebJobs.Script/Host/ScriptHostManager.cs (hunk 6 above)
Comment (brettsam) [reply src/WebJobs.Script/Host/ScriptHostManager.cs]: That being said, I don't think this line is required anyway. I added it while testing something unrelated (which is why the `State = ScriptHostState.Default;` is added below as well). I'll pull them out and test it to make sure.

File: test/WebJobs.Script.Tests.Integration/TestScripts/Node/Scenarios/index.js (hunk 3 above)
Comment (brettsam) [reply test/WebJobs.Script.Tests.Integration/TestScripts/Node/Scenarios/index.js]: Looks like they do pass. Not sure why I did that :-)

File: test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsTestFixture.cs (hunk 2 above)
Comment (brettsam) [reply test/WebJobs.Script.Tests.Integration/ApplicationInsights/ApplicationInsightsTestFixture.cs]: Done.

File: test/WebJobs.Script.Tests/Description/FunctionInvokerBaseTests.cs [hunk 7]
```diff
@@ -133,11 +133,14 @@ public void LogInvocationMetrics_EmitsExpectedEvents()
...
-            int invocation2Duration = (int.Parse(completionEvents[0].Groups["duration"].Value) / 100) * 100;
+
+            int invocation2RawDuration = int.Parse(completionEvents[0].Groups["duration"].Value); // save this for better output below.
+            int invocation2Duration = (invocation2RawDuration / 100) * 100;
 
             Assert.NotEqual(invocation1Duration, invocation2Duration);
             Assert.Equal(2000, invocation1Duration);
-            Assert.Equal(500, invocation2Duration);
+
+            Assert.True(invocation2Duration == 500, $"Expected 500. Actual: {invocation2Duration}. Raw value: {invocation2RawDuration}");
```
Comment (brettsam) [inline test/WebJobs.Script.Tests/Description/FunctionInvokerBaseTests.cs]: Note: I've been seeing this fail quite often so I added some additional logging. This is unrelated to the change.
